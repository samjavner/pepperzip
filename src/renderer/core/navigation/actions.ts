import { createStandardAction } from "typesafe-actions";
import {
    Area,
    CalendarPage,
    HomePage,
    RecipesPage,
    ReferencePage,
    SettingsPage,
    ShoppingPage,
    ToolsPage,
} from "./model";

export const setActiveArea = createStandardAction("navigation/SET_ACTIVE_AREA")<
    Area
>();

export const setActiveHomePage = createStandardAction(
    "navigation/SET_ACTIVE_HOME_PAGE"
)<HomePage>();

export const setActiveRecipesPage = createStandardAction(
    "navigation/SET_ACTIVE_RECIPES_PAGE"
)<RecipesPage>();

export const setActiveCalendarPage = createStandardAction(
    "navigation/SET_ACTIVE_CALENDAR_PAGE"
)<CalendarPage>();

export const setActiveShoppingPage = createStandardAction(
    "navigation/SET_ACTIVE_SHOPPING_PAGE"
)<ShoppingPage>();

export const setActiveReferencePage = createStandardAction(
    "navigation/SET_ACTIVE_REFERENCE_PAGE"
)<ReferencePage>();

export const setActiveToolsPage = createStandardAction(
    "navigation/SET_ACTIVE_TOOLS_PAGE"
)<ToolsPage>();

export const setActiveSettingsPage = createStandardAction(
    "navigation/SET_ACTIVE_SETTINGS_PAGE"
)<SettingsPage>();
