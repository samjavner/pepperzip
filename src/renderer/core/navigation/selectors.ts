import { globalizeSelectors } from "../util";
import { moduleName, State } from "./model";

export const getActiveArea = (state: State) => state.activeArea;

export const getActiveHomePage = (state: State) => state.activeHomePage;

export const getActiveRecipesPage = (state: State) => state.activeRecipesPage;

export const getActiveCalendarPage = (state: State) => state.activeCalendarPage;

export const getActiveShoppingPage = (state: State) => state.activeShoppingPage;

export const getActiveReferencePage = (state: State) =>
    state.activeReferencePage;

export const getActiveToolsPage = (state: State) => state.activeToolsPage;

export const getActiveSettingsPage = (state: State) => state.activeSettingsPage;

const selectors = {
    getActiveArea,
    getActiveHomePage,
    getActiveRecipesPage,
    getActiveCalendarPage,
    getActiveShoppingPage,
    getActiveReferencePage,
    getActiveToolsPage,
    getActiveSettingsPage,
};

export default globalizeSelectors<State, typeof selectors>(
    moduleName,
    selectors
);
