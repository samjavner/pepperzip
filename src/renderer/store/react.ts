import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
