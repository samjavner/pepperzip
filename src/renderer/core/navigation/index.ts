import * as actions from "./actions";
import {
    Area,
    CalendarPage,
    HomePage,
    moduleName,
    RecipesPage,
    ReferencePage,
    SettingsPage,
    ShoppingPage,
    ToolsPage,
} from "./model";
import { reducer } from "./reducer";
import selectors from "./selectors";

export { actions, moduleName, reducer, selectors };

export type Area = Area;
export type CalendarPage = CalendarPage;
export type HomePage = HomePage;
export type RecipesPage = RecipesPage;
export type ReferencePage = ReferencePage;
export type SettingsPage = SettingsPage;
export type ShoppingPage = ShoppingPage;
export type ToolsPage = ToolsPage;
