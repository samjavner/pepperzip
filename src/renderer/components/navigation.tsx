import { IconDefinition } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import classNames from "classnames";
import * as React from "react";

export interface NavigationLayoutProps {
    navbar: NavbarProps;
    tabs: TabProps[];
    page: React.ReactNode;
}

export interface NavbarProps {
    left: NavbarLinkProps[];
    right: NavbarLinkProps[];
}

export interface NavbarLinkProps {
    isActive: boolean;
    onClick: () => void;
    text?: string;
    icon?: IconDefinition;
}

export interface TabsProps {
    tabs: TabProps[];
}

export interface TabProps {
    icon: IconDefinition;
    isActive: boolean;
    onClick: () => void;
    text: string;
}

export const NavigationLayout: React.FunctionComponent<
    NavigationLayoutProps
> = ({ navbar, tabs, page }) => (
    <div
        style={{
            height: "100vh",
            width: "100vw",
            display: "grid",
            gridTemplateRows: "auto 1fr",
        }}
    >
        <div style={{ gridRow: 1 }}>
            <Navbar {...navbar} />
        </div>
        <div style={{ gridRow: 2, overflow: "hidden" }}>
            <div
                style={{
                    height: "100%",
                    display: "grid",
                    gridTemplateRows: "auto 1fr",
                }}
            >
                <div style={{ gridRow: 1 }}>
                    <Tabs tabs={tabs} />
                </div>
                <div style={{ gridRow: 2, overflow: "hidden" }}>{page}</div>
            </div>
        </div>
    </div>
);

const Navbar: React.FunctionComponent<NavbarProps> = ({ left, right }) => (
    <nav className="navbar is-primary">
        <div className="navbar-brand">
            {left.map((props, index) => (
                <NavbarLink key={index} {...props} />
            ))}
        </div>
        <div className="navbar-menu">
            <div className="navbar-end">
                {right.map((props, index) => (
                    <NavbarLink key={index} {...props} />
                ))}
            </div>
        </div>
    </nav>
);

const NavbarLink: React.FunctionComponent<NavbarLinkProps> = ({
    isActive,
    onClick,
    text,
    icon,
}) => (
    <a
        href="#"
        className={classNames({
            "navbar-item": true,
            "is-active": isActive,
        })}
        onClick={onClick}
    >
        {icon && <FontAwesomeIcon fixedWidth={true} icon={icon} />}
        {text}
    </a>
);

const Tabs: React.FunctionComponent<TabsProps> = ({ tabs }) => (
    <div className="tabs is-centered" style={{ marginBottom: "0px" }}>
        <ul>
            {tabs.map((tab, index) => (
                <Tab key={index} {...tab} />
            ))}
        </ul>
    </div>
);

const Tab: React.FunctionComponent<TabProps> = ({
    icon,
    isActive,
    onClick,
    text,
}) => (
    <li
        className={classNames({
            "is-active": isActive,
        })}
    >
        <a href="#" onClick={onClick}>
            <span className="icon is-small">
                <FontAwesomeIcon icon={icon} />
            </span>
            {text}
        </a>
    </li>
);
