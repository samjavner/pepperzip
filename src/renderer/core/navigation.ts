import { Store, UseStore } from "../store";

export type Navigation = Store<Model, Update>;

export function useNavigation(useStore: UseStore<Model, Update>): Navigation {
    return useStore({
        init,
        update,
    });
}

// MODEL

export interface Model {
    activeArea: AreaName;
    activeHomePage: HomePageName;
    activeRecipesPage: RecipesPageName;
    activeCalendarPage: CalendarPageName;
    activeShoppingPage: ShoppingPageName;
    activeReferencePage: ReferencePageName;
    activeToolsPage: ToolsPageName;
    activeSettingsPage: SettingsPageName;
}

export type AreaName =
    | "home"
    | "recipes"
    | "calendar"
    | "shopping"
    | "reference"
    | "tools"
    | "settings";

export type HomePageName = "home";

export type RecipesPageName =
    | "library"
    | "recipe_box"
    | "favorites"
    | "tags"
    | "search";

export type ShoppingPageName = "grocery_lists" | "inventory";

export type CalendarPageName = "calendar";

export type ReferencePageName = "dictionary" | "nutrition";

export type ToolsPageName =
    | "calculator"
    | "timer"
    | "import"
    | "export"
    | "statistics";

export type SettingsPageName = "settings" | "about";

export const init: Model = {
    activeArea: "home",
    activeHomePage: "home",
    activeRecipesPage: "recipe_box",
    activeCalendarPage: "calendar",
    activeShoppingPage: "grocery_lists",
    activeReferencePage: "dictionary",
    activeToolsPage: "calculator",
    activeSettingsPage: "settings",
};

// UPDATE

type Update = typeof update;

export const update = {
    setActiveArea(model: Model, area: AreaName): Model {
        return {
            ...model,
            activeArea: area,
        };
    },
    setActiveHomePage(model: Model, page: HomePageName): Model {
        return {
            ...model,
            activeHomePage: page,
        };
    },
    setActiveRecipesPage(model: Model, page: RecipesPageName): Model {
        return {
            ...model,
            activeRecipesPage: page,
        };
    },
    setActiveCalendarPage(model: Model, page: CalendarPageName): Model {
        return {
            ...model,
            activeCalendarPage: page,
        };
    },
    setActiveShoppingPage(model: Model, page: ShoppingPageName): Model {
        return {
            ...model,
            activeShoppingPage: page,
        };
    },
    setActiveReferencePage(model: Model, page: ReferencePageName): Model {
        return {
            ...model,
            activeReferencePage: page,
        };
    },
    setActiveToolsPage(model: Model, page: ToolsPageName): Model {
        return {
            ...model,
            activeToolsPage: page,
        };
    },
    setActiveSettingsPage(model: Model, page: SettingsPageName): Model {
        return {
            ...model,
            activeSettingsPage: page,
        };
    },
};
