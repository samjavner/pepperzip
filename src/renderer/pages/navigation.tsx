import {
    faArchive,
    faBook,
    faBox,
    faCalculator,
    faCalendar,
    faChartPie,
    faCog,
    faFileExport,
    faFileImport,
    faHeart,
    faHeartbeat,
    faHome,
    faInfoCircle,
    faSearch,
    faShoppingCart,
    faStopwatch,
    faTags,
} from "@fortawesome/free-solid-svg-icons";
import * as React from "react";
import {
    NavbarProps,
    NavigationLayout,
    TabProps,
} from "../components/navigation";
import * as calendar from "./calendar/calendar";
import * as home from "./home/home";
import * as favorites from "./recipes/favorites";
import * as library from "./recipes/library";
import * as recipeBox from "./recipes/recipeBox";
import * as search from "./recipes/search";
import * as tags from "./recipes/tags";
import * as dictionary from "./reference/dictionary";
import * as nutrition from "./reference/nutrition";
import * as about from "./settings/about";
import * as settings from "./settings/settings";
import * as groceryLists from "./shopping/groceryLists";
import * as inventory from "./shopping/inventory";
import * as calculator from "./tools/calculator";
import * as toolsExport from "./tools/export";
import * as toolsImport from "./tools/import";
import * as statistics from "./tools/statistics";
import * as timer from "./tools/timer";
import * as util from "./util";

// MODEL

export interface Model {
    activeArea: Area;
    activeHomePage: HomePage;
    activeRecipesPage: RecipesPage;
    activeCalendarPage: CalendarPage;
    activeShoppingPage: ShoppingPage;
    activeReferencePage: ReferencePage;
    activeToolsPage: ToolsPage;
    activeSettingsPage: SettingsPage;
}

export type Area =
    | "home"
    | "recipes"
    | "calendar"
    | "shopping"
    | "reference"
    | "tools"
    | "settings";

export type HomePage = "home";

export type RecipesPage =
    | "library"
    | "recipe_box"
    | "favorites"
    | "tags"
    | "search";

export type ShoppingPage = "grocery_lists" | "inventory";

export type CalendarPage = "calendar";

export type ReferencePage = "dictionary" | "nutrition";

export type ToolsPage =
    | "calculator"
    | "timer"
    | "import"
    | "export"
    | "statistics";

export type SettingsPage = "settings" | "about";

export const init: Model = {
    activeArea: "home",
    activeHomePage: "home",
    activeRecipesPage: "recipe_box",
    activeCalendarPage: "calendar",
    activeShoppingPage: "grocery_lists",
    activeReferencePage: "dictionary",
    activeToolsPage: "calculator",
    activeSettingsPage: "settings",
};

// UPDATE

export const update = {
    setActiveArea(model: Model, area: Area): Model {
        return {
            ...model,
            activeArea: area,
        };
    },
    setActiveHomePage(model: Model, page: HomePage): Model {
        return {
            ...model,
            activeHomePage: page,
        };
    },
    setActiveRecipesPage(model: Model, page: RecipesPage): Model {
        return {
            ...model,
            activeRecipesPage: page,
        };
    },
    setActiveCalendarPage(model: Model, page: CalendarPage): Model {
        return {
            ...model,
            activeCalendarPage: page,
        };
    },
    setActiveShoppingPage(model: Model, page: ShoppingPage): Model {
        return {
            ...model,
            activeShoppingPage: page,
        };
    },
    setActiveReferencePage(model: Model, page: ReferencePage): Model {
        return {
            ...model,
            activeReferencePage: page,
        };
    },
    setActiveToolsPage(model: Model, page: ToolsPage): Model {
        return {
            ...model,
            activeToolsPage: page,
        };
    },
    setActiveSettingsPage(model: Model, page: SettingsPage): Model {
        return {
            ...model,
            activeSettingsPage: page,
        };
    },
};

// VIEW

export type ViewModel = util.ViewModel<Model, typeof update>;

export interface Props {
    vm: ViewModel;
    recipeBoxVm: recipeBox.ViewModel;
    statisticsVm: statistics.ViewModel;
}

export const Navigation: React.FunctionComponent<Props> = props => (
    <NavigationLayout
        navbar={getNavbar(props.vm)}
        tabs={getTabs(props.vm)}
        page={getPage(props)}
    />
);

// NAVBAR

function getNavbar({ activeArea, setActiveArea }: ViewModel): NavbarProps {
    return {
        left: [
            {
                isActive: activeArea === "home",
                onClick: () => setActiveArea("home"),
                icon: faHome,
            },
            {
                isActive: activeArea === "recipes",
                onClick: () => setActiveArea("recipes"),
                text: "Recipes",
            },
            {
                isActive: activeArea === "calendar",
                onClick: () => setActiveArea("calendar"),
                text: "Calendar",
            },
            {
                isActive: activeArea === "shopping",
                onClick: () => setActiveArea("shopping"),
                text: "Shopping",
            },
            {
                isActive: activeArea === "reference",
                onClick: () => setActiveArea("reference"),
                text: "Reference",
            },
            {
                isActive: activeArea === "tools",
                onClick: () => setActiveArea("tools"),
                text: "Tools",
            },
        ],
        right: [
            {
                isActive: activeArea === "settings",
                onClick: () => setActiveArea("settings"),
                icon: faCog,
            },
        ],
    };
}

// TABS

function getTabs(vm: ViewModel): TabProps[] {
    switch (vm.activeArea) {
        case "home":
            return getHomeTabs(vm.activeHomePage, vm.setActiveHomePage);
        case "recipes":
            return getRecipesTabs(
                vm.activeRecipesPage,
                vm.setActiveRecipesPage
            );
        case "calendar":
            return getCalendarTabs(
                vm.activeCalendarPage,
                vm.setActiveCalendarPage
            );
        case "shopping":
            return getShoppingTabs(
                vm.activeShoppingPage,
                vm.setActiveShoppingPage
            );
        case "reference":
            return getReferenceTabs(
                vm.activeReferencePage,
                vm.setActiveReferencePage
            );
        case "tools":
            return getToolsTabs(vm.activeToolsPage, vm.setActiveToolsPage);
        case "settings":
            return getSettingsTabs(
                vm.activeSettingsPage,
                vm.setActiveSettingsPage
            );
    }
}

function getHomeTabs(
    activePage: HomePage,
    setActivePage: (page: HomePage) => void
): TabProps[] {
    return [
        {
            isActive: activePage === "home",
            onClick: () => setActivePage("home"),
            icon: faHome,
            text: "Home",
        },
    ];
}

function getRecipesTabs(
    activePage: RecipesPage,
    setActivePage: (page: RecipesPage) => void
): TabProps[] {
    return [
        {
            isActive: activePage === "recipe_box",
            onClick: () => setActivePage("recipe_box"),
            icon: faArchive,
            text: "Recipe Box",
        },
        {
            isActive: activePage === "library",
            onClick: () => setActivePage("library"),
            icon: faBook,
            text: "Library",
        },
        {
            isActive: activePage === "favorites",
            onClick: () => setActivePage("favorites"),
            icon: faHeart,
            text: "Favorites",
        },
        {
            isActive: activePage === "tags",
            onClick: () => setActivePage("tags"),
            icon: faTags,
            text: "Tags",
        },
        {
            isActive: activePage === "search",
            onClick: () => setActivePage("search"),
            icon: faSearch,
            text: "Search",
        },
    ];
}

function getCalendarTabs(
    activePage: CalendarPage,
    setActivePage: (page: CalendarPage) => void
): TabProps[] {
    return [
        {
            isActive: activePage === "calendar",
            onClick: () => setActivePage("calendar"),
            icon: faCalendar,
            text: "Calendar",
        },
    ];
}

function getShoppingTabs(
    activePage: ShoppingPage,
    setActivePage: (page: ShoppingPage) => void
): TabProps[] {
    return [
        {
            isActive: activePage === "grocery_lists",
            onClick: () => setActivePage("grocery_lists"),
            icon: faShoppingCart,
            text: "Grocery Lists",
        },
        {
            isActive: activePage === "inventory",
            onClick: () => setActivePage("inventory"),
            icon: faBox,
            text: "Inventory",
        },
    ];
}

function getReferenceTabs(
    activePage: ReferencePage,
    setActivePage: (page: ReferencePage) => void
): TabProps[] {
    return [
        {
            isActive: activePage === "dictionary",
            onClick: () => setActivePage("dictionary"),
            icon: faBook,
            text: "Dictionary",
        },
        {
            isActive: activePage === "nutrition",
            onClick: () => setActivePage("nutrition"),
            icon: faHeartbeat,
            text: "Nutrition",
        },
    ];
}

function getToolsTabs(
    activePage: ToolsPage,
    setActivePage: (page: ToolsPage) => void
): TabProps[] {
    return [
        {
            isActive: activePage === "calculator",
            onClick: () => setActivePage("calculator"),
            icon: faCalculator,
            text: "Calculator",
        },
        {
            isActive: activePage === "timer",
            onClick: () => setActivePage("timer"),
            icon: faStopwatch,
            text: "Timer",
        },
        {
            isActive: activePage === "import",
            onClick: () => setActivePage("import"),
            icon: faFileImport,
            text: "Import",
        },
        {
            isActive: activePage === "export",
            onClick: () => setActivePage("export"),
            icon: faFileExport,
            text: "Export",
        },
        {
            isActive: activePage === "statistics",
            onClick: () => setActivePage("statistics"),
            icon: faChartPie,
            text: "Statistics",
        },
    ];
}

function getSettingsTabs(
    activePage: SettingsPage,
    setActivePage: (page: SettingsPage) => void
): TabProps[] {
    return [
        {
            isActive: activePage === "settings",
            onClick: () => setActivePage("settings"),
            icon: faCog,
            text: "Settings",
        },
        {
            isActive: activePage === "about",
            onClick: () => setActivePage("about"),
            icon: faInfoCircle,
            text: "About",
        },
    ];
}

// PAGE

function getPage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeArea) {
        case "home":
            return getHomePage(props);
        case "recipes":
            return getRecipesPage(props);
        case "calendar":
            return getCalendarPage(props);
        case "shopping":
            return getShoppingPage(props);
        case "reference":
            return getReferencePage(props);
        case "tools":
            return getToolsPage(props);
        case "settings":
            return getSettingsPage(props);
    }
}

function getHomePage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeHomePage) {
        case "home":
            return <home.Page />;
    }
}

function getRecipesPage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeRecipesPage) {
        case "recipe_box":
            return (
                <recipeBox.Page
                    vm={props.recipeBoxVm}
                    statisticsVm={props.statisticsVm}
                />
            );
        case "library":
            return <library.Page />;
        case "favorites":
            return <favorites.Page />;
        case "tags":
            return <tags.Page />;
        case "search":
            return <search.Page />;
    }
}

function getCalendarPage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeCalendarPage) {
        case "calendar":
            return <calendar.Page />;
    }
}

function getShoppingPage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeShoppingPage) {
        case "grocery_lists":
            return <groceryLists.Page />;
        case "inventory":
            return <inventory.Page />;
    }
}

function getReferencePage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeReferencePage) {
        case "dictionary":
            return <dictionary.Page />;
        case "nutrition":
            return <nutrition.Page />;
    }
}

function getToolsPage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeToolsPage) {
        case "calculator":
            return <calculator.Page />;
        case "timer":
            return <timer.Page />;
        case "import":
            return <toolsImport.Page />;
        case "export":
            return <toolsExport.Page />;
        case "statistics":
            return <statistics.Page {...props.statisticsVm} />;
    }
}

function getSettingsPage(props: Props): React.ReactElement<any> {
    switch (props.vm.activeSettingsPage) {
        case "settings":
            return <settings.Page />;
        case "about":
            return <about.Page />;
    }
}
