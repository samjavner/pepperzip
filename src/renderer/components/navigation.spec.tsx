import { faBox, faCog, faHome } from "@fortawesome/free-solid-svg-icons";
import * as React from "react";
import * as renderer from "react-test-renderer";
import { NavigationLayout, NavigationLayoutProps } from "./navigation";

describe("navigation components", () => {
    describe("NavigationLayout", () => {
        let props: NavigationLayoutProps;

        beforeEach(() => {
            props = {
                navbar: {
                    left: [
                        {
                            isActive: false,
                            icon: faHome,
                            onClick: () => ({}),
                        },
                        {
                            isActive: false,
                            text: "abc",
                            onClick: () => ({}),
                        },
                    ],
                    right: [
                        {
                            isActive: true,
                            text: "xyz",
                            onClick: () => ({}),
                        },
                        {
                            isActive: false,
                            icon: faCog,
                            onClick: () => ({}),
                        },
                    ],
                },
                tabs: [
                    {
                        isActive: false,
                        text: "ABC",
                        icon: faCog,
                        onClick: () => ({}),
                    },
                    {
                        isActive: true,
                        text: "XYZ",
                        icon: faBox,
                        onClick: () => ({}),
                    },
                ],
                page: "Hello!",
            };
        });

        it("should have expected layout", () => {
            const component = renderer.create(
                <NavigationLayout
                    navbar={{ left: [], right: [] }}
                    tabs={[]}
                    page={null}
                />
            );
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });

        it("should have expected navbar, tabs, and page", () => {
            const component = renderer.create(<NavigationLayout {...props} />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });

        it("handles clicking on navbar links", () => {
            let clicked = false;
            props.navbar.right[0].onClick = () => {
                clicked = true;
            };
            const component = renderer.create(<NavigationLayout {...props} />);
            const link = component.root.find(
                node =>
                    node.type === "a" &&
                    node.props.className &&
                    node.props.className.indexOf("is-active") !== -1
            );
            link.props.onClick();
            expect(clicked).toBe(true);
        });

        it("handles clicking on tabs", () => {
            let clicked = false;
            props.tabs[1].onClick = () => {
                clicked = true;
            };
            const component = renderer.create(<NavigationLayout {...props} />);
            const link = component.root.find(
                node =>
                    node.type === "li" &&
                    node.props.className &&
                    node.props.className.indexOf("is-active") !== -1
            );
            (link.children[0] as renderer.ReactTestInstance).props.onClick();
            expect(clicked).toBe(true);
        });
    });
});
