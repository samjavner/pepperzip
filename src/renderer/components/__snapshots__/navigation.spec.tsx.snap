// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation components NavigationLayout should have expected layout 1`] = `
<div
  style={
    Object {
      "display": "grid",
      "gridTemplateRows": "auto 1fr",
      "height": "100vh",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "gridRow": 1,
      }
    }
  >
    <nav
      className="navbar is-primary"
    >
      <div
        className="navbar-brand"
      />
      <div
        className="navbar-menu"
      >
        <div
          className="navbar-end"
        />
      </div>
    </nav>
  </div>
  <div
    style={
      Object {
        "gridRow": 2,
        "overflow": "hidden",
      }
    }
  >
    <div
      style={
        Object {
          "display": "grid",
          "gridTemplateRows": "auto 1fr",
          "height": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "gridRow": 1,
          }
        }
      >
        <div
          className="tabs is-centered"
          style={
            Object {
              "marginBottom": "0px",
            }
          }
        >
          <ul />
        </div>
      </div>
      <div
        style={
          Object {
            "gridRow": 2,
            "marginTop": "-3px",
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "height": "100%",
              "paddingTop": "3px",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`navigation components NavigationLayout should have expected navbar, tabs, and page 1`] = `
<div
  style={
    Object {
      "display": "grid",
      "gridTemplateRows": "auto 1fr",
      "height": "100vh",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "gridRow": 1,
      }
    }
  >
    <nav
      className="navbar is-primary"
    >
      <div
        className="navbar-brand"
      >
        <a
          className="navbar-item"
          href="#"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-home fa-w-18 fa-fw "
            data-icon="home"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </a>
        <a
          className="navbar-item"
          href="#"
          onClick={[Function]}
        >
          abc
        </a>
      </div>
      <div
        className="navbar-menu"
      >
        <div
          className="navbar-end"
        >
          <a
            className="navbar-item is-active"
            href="#"
            onClick={[Function]}
          >
            xyz
          </a>
          <a
            className="navbar-item"
            href="#"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-cog fa-w-16 fa-fw "
              data-icon="cog"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </a>
        </div>
      </div>
    </nav>
  </div>
  <div
    style={
      Object {
        "gridRow": 2,
        "overflow": "hidden",
      }
    }
  >
    <div
      style={
        Object {
          "display": "grid",
          "gridTemplateRows": "auto 1fr",
          "height": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "gridRow": 1,
          }
        }
      >
        <div
          className="tabs is-centered"
          style={
            Object {
              "marginBottom": "0px",
            }
          }
        >
          <ul>
            <li
              className=""
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <span
                  className="icon is-small"
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-cog fa-w-16 "
                    data-icon="cog"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </span>
                ABC
              </a>
            </li>
            <li
              className="is-active"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <span
                  className="icon is-small"
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-box fa-w-16 "
                    data-icon="box"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M509.5 184.6L458.9 32.8C452.4 13.2 434.1 0 413.4 0H272v192h238.7c-.4-2.5-.4-5-1.2-7.4zM240 0H98.6c-20.7 0-39 13.2-45.5 32.8L2.5 184.6c-.8 2.4-.8 4.9-1.2 7.4H240V0zM0 224v240c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V224H0z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </span>
                XYZ
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        style={
          Object {
            "gridRow": 2,
            "marginTop": "-3px",
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "height": "100%",
              "paddingTop": "3px",
            }
          }
        >
          Hello!
        </div>
      </div>
    </div>
  </div>
</div>
`;
